### ISSUE TITLE
what batch_masked_embedding_lookup exactly do?

### ISSUE THREAD

**Issue by fword** [chenghuige/hasky#4]
when i used to do nlp jobs, i just use embedding_lookup. but when i read your code. find batch_masked_embedding_lookup  is very confusing, what does it exactly do? does it work better than embedding_lookup in ur experiment?

**Comment by chenghuige** [comment]
It is just a wrapper of embedding_lookup, adding choice of sum or mean operation after embedding_lookup, also if you want to not consider not to include index 0's embedding(assume this index with all 0 embedding)  you can set exclude_zero_index=True. Why exclude 0's index, because 0 index might means padding.

**Comment by fword** [comment]
exclude_zero_index is ez understant, so what is this code below means
def reduce_mean_with_mask(input_tensor, mask, reduction_indices=None, keep_dims=False):
  return  tf.reduce_sum(input_tensor, reduction_indices=reduction_indices, keep_dims=keep_dims) / \
          tf.reduce_sum(mask, reduction_indices=reduction_indices, keep_dims=keep_dims)

**Comment by fword** [comment]
aother little question:

if i set FLAGS.num_gpus=0, does it means it will use one gpu to train? but when i run 'nvidia-smi' .it seems four gpus are running on my code. so do i why need this write code like:  loss = melt.tower_losses(loss_function, FLAGS.num_gpus) to explicitly  to use the gpu?

**Comment by chenghuige** [comment]
reduce_mean_with_mask means adding all embedding then divide by num embeddings not masked(zero padded). 
For num_gpu = 0 it is single gpu concern,however tf will try to grab all gpus for any tf program though not much speed improve..You can change with setting CUDA visible gpu env to force using only one gpu. If you set num_gpu=2 batch_size=128 that means you run two gpus each with 128 size data(256 batch size each step) might improve speed but data transfer might also decease perf.

**Comment by fword** [comment]
thanks a lot, this is very helpful. then close this issue
